### PR TITLE
Add push_vote failure path

### LINKED ISSUE #16680: cluster_info::test_push_votes_with_tower is flaky
#### Problem

Running the following command fails in about 8/10 of the runs for me:

```
$ cargo test --package solana-core --release test_push_votes_with_tower
    Finished release [optimized] target(s) in 0.50s
     Running target/release/deps/solana_core-3de16058cf8e4670

running 1 test
test cluster_info::tests::test_push_votes_with_tower ... FAILED

failures:

---- cluster_info::tests::test_push_votes_with_tower stdout ----
thread 'cluster_info::tests::test_push_votes_with_tower' panicked at 'assertion failed: `(left == right)`
  left: `28`,
 right: `31`', core/src/cluster_info.rs:4065:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    cluster_info::tests::test_push_votes_with_tower
```

The left value varies across runs. I can reproduce with both c8b474cd0ba2a66cc948b92149d747a7438d672b and v1.5.18. Without `--release` it does not reproduce.

#### Proposed Solution

I know very little about Solana, but looking at the test itself, there are several sources of nondeterminism, such as the `ThreadRng`, getting the current time, and probably also `KeyPair::new()`? I tried replacing the `ThreadRng` with one with a fixed seed, but that did not solve the problem. Replacing `timestamp()` with `0` makes it pass more often, but then it sometimes fails on a different assertion:

```
thread 'cluster_info::tests::test_push_votes_with_tower' panicked at 'assertion failed: vote_slot != 23', core/src/cluster_info.rs:4079:13
```

### PR DESCRIPTION
**Pull request by xiangzhu70**
push_vote may fail due to wallclock checks.  The current
function prototype lacks return value, likely to mislead
people to believe that push_vote always succeed in pushing
the vote into the datastore.
test_push_vote_with_tower failed due to this misled
assumption.
Adding the error return and handling the error fixes this
test.  Other push_vote calls in the code are appended with
unwrap_or_default() so the behavior is kept the same, but
error ignoring is more apparent.

#### Summary of Changes

Fixes #16680

### CODE REVIEW

**Comment by behzadnouri** [comment]
What problem is this solving?
The flaky test was fixed in https://github.com/solana-labs/solana/pull/26554

The test was flaky because it was creating an unrealistic scenario (voting for 32 slots in the same milliseconds) which the actual running validator does not encounter. We don't want to add more complexity to the code for solving problems we don't have.

**Comment by xiangzhu70** [comment]
If it is a totally unrealistic scenario, why is there even logic to handle that case (comparing hash and returning false)?  there should be panic there.

It is not clean to do some case handling on some unlikely case, returning false at the lower level, then silently ignoring the error return at the higher level.  Then the higher level tests call the function and got unexpected result which is exactly the reason of this test failure.  What this PR does is to make the function returning behavior more explicit, and forcing the higher level callers to be aware of possible error returns.

File: core/src/voting_service.rs [hunk 1]
```diff
@@ -95,7 +95,7 @@ impl VotingService {
             VoteOp::PushVote {
                 tx, tower_slots, ..
             } => {
-                cluster_info.push_vote(&tower_slots, tx);
+                cluster_info.push_vote(&tower_slots, tx).unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (xiangzhu70) [reply core/src/voting_service.rs:98]: Here the intention is just to maintain the existing behavior, which is to silently ignore the error inside this function.  unwrap() will cause panic when it fails.  unwrap_or_default maintains the existing behavior.

File: gossip/src/cluster_info.rs [hunk 2]
```diff
@@ -3874,15 +3889,29 @@ RPC Enabled Nodes: 1"#;
...
-        // New versioned-crds-value should have wallclock later than existing
-        // entries, otherwise might not get inserted into the table.
-        sleep(Duration::from_millis(5));
-        cluster_info.push_vote(&tower, vote);
+        let mut slot_not_existing: Slot = 23;
+        let mut push_vote_failed = false;
+        if let Err(_err) = cluster_info.push_vote(&tower, vote.clone()) {
+            slot_not_existing = MAX_LOCKOUT_HISTORY as Slot;
+            push_vote_failed = true;
+        }
```
[the remark this answers is not in the document]
Comment (xiangzhu70) [reply gossip/src/cluster_info.rs:3897]: Agreed.  But I don't know yet how to force the wallclock to be the same.  Let me look into it.

**Comment by behzadnouri** [comment]
> If it is a totally unrealistic scenario, why is there even logic to handle that case (comparing hash and returning false)? there should be panic there.

Gossip can and should be able to ingest any entry regardless of its wallclock.
Sending 32 votes within the same millisecond is the unrealistic part.
You can start a validator and see for yourself that we don't need this code.

**Comment by xiangzhu70** [comment]
> > If it is a totally unrealistic scenario, why is there even logic to handle that case (comparing hash and returning false)? there should be panic there.
> 
> Gossip can and should be able to ingest any entry regardless of its wallclock. Sending 32 votes within the same millisecond is the unrealistic part. You can start a validator and see for yourself that we don't need this code.

If you strongly believe it is unrealistic, then in crds.rs:overrides() "Ordering::Equal => " case it should not do hash comparison and return a choice.  It should just call panic().  That would be a more deterministic design, not leaving room for people to believe push_vote always succeed, write test cases against it, and got flaky test results.  Any such test would run into that panic immediately, instead of being a flaky test to be debugged.

**Comment by behzadnouri** [comment]
> If you strongly believe it is unrealistic, then in crds.rs:overrides() "Ordering::Equal => " case it should not do hash comparison and return a choice. It should just call panic().

I already mentioned here: https://github.com/solana-labs/solana/pull/26584#issuecomment-1182051163
> Gossip can and should be able to ingest any entry regardless of its wallclock.

I also mentioned on the original issue: https://github.com/solana-labs/solana/issues/16680#issuecomment-1180525841
> By design, and with the exception of NodeInstance, gossip overrides entries if the new value has a more recent wallclock.
> It is totally valid for gossip to drop inserts if the table already has a more recent entry for the same key.
> That is not an error.

Ties are broken using hash comparison because they should be treated in the exact same deterministic way across the cluster.

**Pull request closed without merging**